ocaml/xapi/repository_helpers.ml

```diff
@@ -1361,15 +1324,139 @@ let prune_updateinfo_for_livepatches livepatches updateinfo =
   in
   {updateinfo with livepatches= lps}
 
-let do_with_device_models ~__context ~host f =
-  (* Call f with device models of all running HVM VMs on the host *)
-  Db.Host.get_resident_VMs ~__context ~self:host
-  |> List.map (fun self -> (self, Db.VM.get_record ~__context ~self))
-  |> List.filter (fun (_, record) -> not record.API.vM_is_control_domain)
-  |> List.filter_map f
-  |> function
-  | [] ->
-      ()
-  | _ :: _ ->
-      let host' = Ref.string_of host in
-      raise Api_errors.(Server_error (cannot_restart_device_model, [host']))
+let do_with_host_pending_guidances ~op guidances =
+  List.iter
+    (fun g ->
+      match Guidance.to_pending_guidance g with
+      | ( Some `restart_toolstack
+        | Some `reboot_host
+        | Some `reboot_host_on_xen_livepatch_failure
+        | Some `reboot_host_on_kernel_livepatch_failure
+        | Some `reboot_host_on_livepatch_failure ) as g' ->
+          Option.iter op g'
+      | _ ->
+          ()
+    )
+    guidances
+
+let do_with_vm_pending_guidances ~op ~vm guidances =
+  List.iter
+    (fun g ->
+      match Guidance.to_pending_guidance g with
+      | Some `restart_device_model ->
+          op vm `restart_device_model
+      | Some `restart_vm ->
+          op vm `restart_vm
+      | _ ->
+          ()
+    )
+    guidances
+
+let merge_with_pending_guidances ~pending ~coming =
+  let open GuidanceSet in
+  let pending = of_list pending in
+  union pending (of_list coming) |> remove EvacuateHost |> fun unioned ->
+  (diff unioned pending |> elements, diff pending unioned |> elements)
+
+let is_livepatch_failure = function
+  | Guidance.RebootHostOnLivePatchFailure
+  | Guidance.RebootHostOnXenLivePatchFailure
+  | Guidance.RebootHostOnKernelLivePatchFailure ->
+      true
+  | _ ->
+      false
+
+type pending_ops = {
+    host_get: unit -> Guidance.t list
+  ; host_add:
+         [ `reboot_host
+         | `reboot_host_on_livepatch_failure
+         | `restart_toolstack
+         | `reboot_host_on_xen_livepatch_failure
+         | `reboot_host_on_kernel_livepatch_failure ]
+      -> unit
+  ; host_remove:
+         [ `reboot_host
+         | `reboot_host_on_livepatch_failure
+         | `restart_toolstack
+         | `reboot_host_on_xen_livepatch_failure
+         | `reboot_host_on_kernel_livepatch_failure ]
+      -> unit
+  ; vms_get: unit -> (string * Guidance.t list) list
+  ; vm_add: string -> [`restart_device_model | `restart_vm] -> unit
+  ; vm_remove: string -> [`restart_device_model | `restart_vm] -> unit
+}
+
+let get_ops_of_pending ~__context ~host ~kind =
+  let get_pending_guidances_of_host ~get =
+    get ~__context ~self:host
+    |> List.map (fun g -> Guidance.of_pending_guidance g)
+  in
```

Comment:
Seems we can use `Db.Host.get_pending_guidances` directly inside the function without passing it as parameter `get`?